vertical-pod-autoscaler/recommender/cluster/cluster_feeder.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/golang/glog"
+	vpa_api "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned/typed/poc.autoscaling.k8s.io/v1alpha1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/recommender/model"
+)
+
+// ClusterStateFeeder can update state of ClusterState object.
+type ClusterStateFeeder interface {
+	// Subscribe saves a pointer to a ClusterState object which will be updated.
+	Subscribe(cluster *model.ClusterState)
+	// Feed uptades state of subscribed ClusterState
+	Feed()
+}
```

Comment:
I have some concerns regarding this design (Feeder).
As far as I understand this API enforces synchronized access to all data sources (metrics, pods, vpas, probably events in future) at the same time (as opposed to allowing the recommender to query the data clients with different frequencies). Is this a conscious design decision?
How often is Feed() supposed to be called? Let's assume that new usage samples appear once per minute, but we want to learn about changes to pod specs and VPAs in < 5 sec.
Are we going to call Feed() every 5 sec? What will the MetricsClient do in such case?
My understanding is that a call to Feed() doesn't trigger any actual queries to the underlying data sources, the data clients fetch the data asynchronously, right? Are they responsible for caching the recent data until the next call to Feed()?
Thinking about it, I'm leaning towards explicitly accessing individual clients in the Recommender. But I'm open for discussion.